src/citrine/_utils/batcher.py

```diff
@@ -0,0 +1,164 @@
+from typing import List
+
+from citrine.resources.data_concepts import DataConcepts
+
+from gemd.util import writable_sort_order
+from gemd.entity.link_by_uid import LinkByUID
+
+
+class Batcher(ABC):
+    """Base class for Data Concepts batching routines."""
+
+    @abstractmethod
+    def batch(self, objects: List[DataConcepts], batch_size) -> List[List[DataConcepts]]:
+        """Collect a list of DataConcepts into batches according to some batching algorithm."""
+
+    @classmethod
+    def by_type(cls):
+        """Return a BatchByType batcher."""
+        return BatchByType()
+
+    @classmethod
+    def by_dependency(cls):
+        """Return a BatchByDependency batcher."""
+        return BatchByDependency()
+
+
+class BatchByType(Batcher):
+    """Batching by object type."""
+
+    def batch(self, objects, batch_size):
+        """Collect object batches by type, following an order that will satisfy prereqs."""
+        batches = list()
+        by_type = defaultdict(list)
+        for obj in objects:
+            by_type[obj.typ].append(obj)
+        typ_groups = sorted(list(by_type.values()), key=lambda x: writable_sort_order(x[0]))
+        for typ_group in typ_groups:
+            num_batches = len(typ_group) // batch_size
+            for batch_num in range(num_batches + 1):
+                batch = typ_group[batch_num * batch_size: (batch_num + 1) * batch_size]
+                batches.append(batch)
+
+        return batches
+
+
+class BatchByDependency(Batcher):
+    """Batching by clusters where nothing references anything outside the cluster."""
+
+    def batch(self, objects, batch_size):
+        """Collect object batches that are internally consistent for dry_run object tests."""
+        from gemd.entity.object.has_parameters import HasParameters
+        from gemd.entity.object.has_conditions import HasConditions
+        from gemd.entity.object.has_properties import HasProperties
+        from gemd.entity.object.has_template import HasTemplate
+        from gemd.entity.template.has_parameter_templates import HasParameterTemplates
+        from gemd.entity.template.has_condition_templates import HasConditionTemplates
+        from gemd.entity.template.has_property_templates import HasPropertyTemplates
+
+        from gemd.entity.object import IngredientRun, MaterialRun, MeasurementRun, ProcessRun
+        from gemd.entity.object import IngredientSpec, MaterialSpec  # no ProcessSpec
+
+        # Collect shallow dependences, UID references, and type-based clusters
+        depends = defaultdict(list)
+        derefs = dict()
+        by_type = defaultdict(list)
+        for obj in objects:
+            derefs[obj] = obj  # no-op to skip conditional later
+            # Index LinkByUIDs
+            for this_scope in obj.uids:
+                derefs[LinkByUID.from_entity(obj, scope=this_scope)] = obj
+
+            # Index attribute templates from attributes
+            if isinstance(obj, HasParameters):
```

Comment:
I think it would make sense to move this dependency generation logic to a separate utility function to avoid such a long method.

I'm okay with it being a method on `BatchByDependency`, or it could be a separate top-level function somewhere in the package.

However, is this logic not elsewhere in in `gemd-python` or `citrine-python` already? This smells very similar to a `flatten` routine, although that is probably a naive first thought on my end.

Reply:
It could reasonably be a gemd-python utility (`all_dependencies` or similar), but it is not a method right now.  I'll abstract it for cleanliness now, and maybe implement similar logic in gemd-python in the future.